server/pulp/server/controllers/importer.py

```diff
@@ -0,0 +1,281 @@
+    result = importer_instance.validate_config(transfer_repo, call_config)
+
+    # For backward compatibility with plugins that don't yet return the tuple
+    if isinstance(result, bool):
+        valid_config = result
+        message = None
+    else:
+        valid_config, message = result
+
+    if not valid_config:
+        raise exceptions.PulpCodedValidationException(validation_errors=message)
+
+
+@celery.task(base=Task, name='pulp.server.managers.repo.importer.remove_importer')
+def remove_importer(repo_id):
+    """
+    Removes an importer from a repository.
+
+    :param repo_id: identifies the repo
+    :type  repo_id: str
+    """
+    repo_obj = model.Repository.objects.get_repo_or_missing_resource(repo_id)
+    repo_importer = model.Importer.objects.get_or_404(repo_id=repo_id)
+
+    # remove schedules
+    sync_manager = manager_factory.repo_sync_schedule_manager()
+    sync_manager.delete_by_importer_id(repo_id, repo_importer.importer_type_id)
+
+    # Call the importer's cleanup method
+    importer_instance, plugin_config = plugin_api.get_importer_by_id(repo_importer.importer_type_id)
+
+    call_config = PluginCallConfiguration(plugin_config, repo_importer.config)
+    transfer_repo = repo_obj.to_transfer_repo()
+    importer_instance.importer_removed(transfer_repo, call_config)
+    repo_importer.delete()
+
+
+def queue_remove_importer(repo_id, importer_type_id):
+    """
+    Dispatch a task to remove the importer from a repository.
+
+    :param repo_id: identifies the repo
+    :type  repo_id: str
+    :param importer_type_id: type of importer
+    :type  importer_type_id: str
+
+    :return: asynchronous result
+    :rtype:  pulp.server.async.tasks.TaskResult
+    """
+    get_valid_importer(repo_id, importer_type_id)
+    task_tags = [tags.resource_tag(tags.RESOURCE_REPOSITORY_TYPE, repo_id),
+                 tags.resource_tag(tags.RESOURCE_REPOSITORY_IMPORTER_TYPE, importer_type_id),
+                 tags.action_tag('delete_importer')]
+    async_result = remove_importer.apply_async_with_reservation(
+        tags.RESOURCE_REPOSITORY_TYPE, repo_id, [repo_id], tags=task_tags)
+    return async_result
+
+
+def get_valid_importer(repo_id, importer_type_id):
+    """
+    Retrieves the importer for the specified repo and ensures that the provided importer type id
+    matches the importer set on the repo.
+
+    :param repo_id: id of the repo
+    :type  repo_id: str
+    :param importer_type_id: type of the importer
+    :type  importer_type_id: str
+
+    :return: key-value pairs describing the importer in use
+    :rtype:  dict
+
+    :raises pulp.server.exceptions.MissingResource: if repo/importer combination cannot be found
+    """
+    model.Repository.objects.get_repo_or_missing_resource(repo_id)
+    try:
+        importer = model.Importer.objects.get_or_404(repo_id=repo_id)
+        if importer.importer_type_id != importer_type_id:
+            raise exceptions.MissingResource(importer_id=importer_type_id)
```

Comment:
This will get caught on the next line, thrown away, and replaced by a new but identical exception. One solution is to move these two lines below the ``except`` block. Then you at least have two distinct code paths that end up raising the same exception.